File: java/tools/src/main/java/io/github/ajoz/workshop/fp/tools/CurriedFunctions.java
package io.github.ajoz.workshop.fp.tools;

@SuppressWarnings("unused")
public final class CurriedFunctions {
    private CurriedFunctions() {
        throw new IllegalStateException("CurriedFunctions should not be instantiated!");
    }

    public static <A, B, C> Function1<B, Function1<A, C>> flip(final Function1<A, Function1<B, C>> function1) {
        return (B b) -> (A a) -> function1.apply(a).apply(b);
    }

    public static <A, B, C> Function1<B, C> applyFirst(final Function1<A, Function1<B, C>> function,
                                                       final A value) {
        return (B b) -> function.apply(value).apply(b);
    }

    public static <A, B, C> Function1<B, C> applyFirst(final Function1<A, Function1<B, C>> function,
                                                       final Supplier<A> supplier) {
        return (B b) -> function.apply(supplier.get()).apply(b);
    }

    public static <A, B, C> Function1<A, C> applySecond(final Function1<A, Function1<B, C>> function,
                                                        final B value) {
        return (A a) -> function.apply(a).apply(value);
    }

    public static <A, B, C> Function1<A, C> applySecond(final Function1<A, Function1<B, C>> function,
                                                        final Supplier<B> supplier) {
        return (A a) -> function.apply(a).apply(supplier.get());
    }
}

File: kotlin/tools/src/main/kotlin/io/github/ajoz/workshop/fp/kotlin/tools/CurriedFunctions.kt
package io.github.ajoz.workshop.fp.kotlin.tools


// applies the first argument of the curried function from a given supplier
fun <A, B, C> applyFirst(function: (A) -> (B) -> C, supplier: () -> A): (B) -> C =
        { b: B ->
            function(supplier())(b)
        }

// applyFirst version that is an infix extension function
infix fun <A, B, C> ((A) -> (B) -> C).apF(supplier: () -> A): (B) -> C =
        applyFirst(this, supplier)

// applies the second argument of the curried function from a given supplier
fun <A, B, C> applySecond(function: (A) -> (B) -> C, supplier: () -> B): (A) -> C =
        { a: A ->
            function(a)(supplier())
        }

// applySecond version that is an infix extension function
infix fun <A, B, C> ((A) -> (B) -> C).apS(supplier: () -> B): (A) -> C =
        applySecond(this, supplier)